src/sudoku.rs

```diff
@@ -0,0 +1,267 @@
+// If some constraint can only be satisfied by 1 entry, it will be entered immediately
+// This is equivalent to finding naked singles and hidden singles
+// If no entry can be deduced, a constraint with the least amount of possibilites
+// is chosen and all possibilites tried out.
+
+// helper struct for recursive solving
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct SudokuSolver {
+	pub grid: Sudoku,
+	pub covers: Covers,
+}
+
+impl SudokuSolver {
+	pub fn from_sudoku(sudoku: Sudoku) -> SudokuSolver {
+		SudokuSolver {
+			covers: Covers::from_sudoku(&sudoku),
+			grid: sudoku,
+		}
+	}
+
+	#[inline]
+	fn remove_impossible(&mut self, new_entry: Entry) {
+		self.covers.remove_impossible(new_entry);
+	}
+
+	fn insert_entry(&mut self, entry: Entry) {
+		self.grid.0[entry.cell()] = entry.num();
+		self.remove_impossible(entry);
+	}
+
+	fn with_entry(&self, entry: Entry) -> Self {
+		let mut sudoku = self.clone();
+		sudoku.insert_entry(entry);
+		sudoku
+	}
+
+	#[inline]
+	pub fn is_solved(&self) -> bool {
+		&self.covers.covered[..] == &[true; 324][..]
+	}
+
+	#[inline]
+	fn is_impossible(&self) -> bool {
+		Iterator::zip( self.covers.possibilities_count.iter(), self.covers.covered.iter() )
+			.any(|(&poss, &covered)| !covered && poss == 0)
+	}
+
+	// return true if new entries were found
+	fn insert_deduced_entries(&mut self) -> bool {
+		let entries = self.covers.possibilities_count.iter()
+			.enumerate()
+			.filter(|&(_, &n_poss)| n_poss == 1)
+			.map(|(idx, _)| self.matching_entry(idx) )//	self.covers.entries.iter().cloned().find(|e| e.constrains(idx))
+			.collect::<Vec<_>>();
+
+		let entries_added = entries.len() != 0;
+		for entry in entries {
+			self.insert_entry(entry);
+		}
+		entries_added
+	}
+
+	// may fail, but only if used incorrectly
+	#[inline]
+	fn matching_entry(&self, constraint_nr: usize) -> Entry {
+		self.covers.entries.iter()
+			.cloned()
+			.find(|e| e.constrains(constraint_nr))
+			.unwrap()
+	}
+
+	pub fn solve_one(self) -> Option<Sudoku> {
+		let result = self.solve_at_most(1);
+		if result.len() == 0 {
+			None
+		} else {
+			result.into_iter().next() // just take one
+		}
+	}
+
+	pub fn solve_unique(self) -> Option<Sudoku> {
+		let result = self.solve_at_most(2);
+		if result.len() == 1 {
+			result.into_iter().next()
+		} else {
+			None
+		}
+	}
+
+	pub fn solve_at_most(self, limit: usize) -> Vec<Sudoku> {
+		let mut solutions = vec![];
+		self._solve_at_most(limit, &mut solutions);
+		solutions
+	}
+
+	fn _solve_at_most(mut self, limit: usize, solutions: &mut Vec<Sudoku>) {
+		if solutions.len() == limit { return }
+
+		// deduce entries, but check in between deductions if the sudoku is still possible
+		while self.insert_deduced_entries() {
+			if self.is_impossible() { return }
+		}
+
+		// impossible to insert another number
+		// either solved or unsolvable
+		if self.covers.is_empty() && self.is_solved() {
+			if self.is_solved() {
```

Comment:
Looks like I wanted to remove the nesting and stopped halfway.
Yeah it is an error. `is_solved()` has no side effects.